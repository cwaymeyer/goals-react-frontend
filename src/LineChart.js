import { useContext, useEffect } from "react";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
  TimeScale,
} from "chart.js";
import { Line } from "react-chartjs-2";
import "chartjs-adapter-date-fns";
import ProgressContext from "./context/progressContext";

function LineChart({ goalData }) {
  const { progressData, setProgressData } = useContext(ProgressContext);

  /** Use progress data to create array of objects to populate chart
   *
   * [ { x: Date, y: 1RM }, { x: Date, y: 1RM } ]
   */
  useEffect(
    function setup() {
      goalData.progress.map((prog) => {
        setProgressData((progressData) => [
          ...progressData,
          { x: +prog.date, y: prog.orm },
        ]);
      });
    },
    [goalData]
  );

  ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    TimeScale,
    Title,
    Tooltip,
    Legend
  );

  const options = {
    responsive: true,
    scales: {
      x: {
        type: "time",
        time: {
          unit: "week",
        },
      },
    },
  };

  const data = {
    datasets: [
      {
        label: "My 1RM",
        data: progressData,
        borderColor: "rgb(255, 99, 132)",
        backgroundColor: "rgba(255, 99, 132, 0.5)",
      },
      {
        label: "Target 1RM",
        data: [
          { x: +goalData.start_date, y: goalData.target_weight },
          { x: +goalData.end_date, y: goalData.target_weight },
        ],
        borderColor: "rgb(53, 162, 235)",
        backgroundColor: "rgba(53, 162, 235, 0.5)",
      },
    ],
  };

  return <Line options={options} data={data} />;
}

export default LineChart;
