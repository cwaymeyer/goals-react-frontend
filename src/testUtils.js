import UserContext from "./context/userContext";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

const progress = [
  {
    id: 10,
    goal_id: 1,
    weight: 50,
    reps: 5,
    orm: "56.3",
    date: "1645477469350",
  },
  {
    id: 11,
    goal_id: 1,
    weight: 50,
    reps: 6,
    orm: "58.1",
    date: "1645477669350",
  },
];

const starting_progress = {
  orm: "56.3",
  date: "1645477269350",
};

const latest_progress = {
  orm: "58.1",
  date: "1645477669350",
};

const goal = {
  id: "1",
  name: "Test Goal",
  username: "demoUser",
  start_weight: null,
  target_weight: 100,
  timeline: 3,
  start_date: "1645477269350",
  end_date: "1653163269350",
  progress: progress,
  starting_progress: starting_progress,
  latest_progress: latest_progress,
};

const demoUser = {
  username: "demoUser",
  password: "password",
  goals: [goal],
};

const UserProvider = ({ children, currentUser = demoUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

const renderWithRouter = (ui, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return render(ui, { wrapper: BrowserRouter });
};

export {
  UserProvider,
  renderWithRouter,
  goal,
  starting_progress,
  latest_progress,
};
