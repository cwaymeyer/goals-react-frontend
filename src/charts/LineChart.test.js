import { render } from "@testing-library/react";
import { goal } from "../testUtils";
import LineChart from "./LineChart";

it("renders without crashing", function () {
  const setProgressData = jest.fn();
  render(<LineChart goalData={goal} setProgressData={setProgressData} />);
});
