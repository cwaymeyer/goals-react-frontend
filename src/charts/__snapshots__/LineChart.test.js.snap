// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <canvas
    height="150"
    role="img"
    width="300"
  />
</DocumentFragment>
`;
